### PR TITLE
fresh branch to introduce cache optimization.

### PR DESCRIPTION
**Pull request by dahliamalkhi**
This PR prevents checkpoint entries from entering the AddressSpaceView cache.

### CODE REVIEW

**Comment by corfudb-performance** [comment]
Results automatically generated by _CorfuDB Benchmark Framework_ to assess the performance of this pull request for commit **7823bd2**.

*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 10 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Iterator workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Iterator workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Iterator workload, 10 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 10 threads, Disk mode
 
 An interactive dashboard with Pull Request Performance Metrics for ALL cluster types and numbers of threads in run, is available at:
 [Pull Request #869 Graphs ](https://snapshot.raintank.io/dashboard/snapshot/6TjjRxkvAzkespjpconYNKPfh2bINwmJ)

**Comment by slfritchie** [comment]
Replaces #814.

File: runtime/src/main/java/org/corfudb/recovery/RecoveryUtils.java [hunk 1]
```diff
@@ -53,18 +52,17 @@ static void createSmrMapIfNotExist(CorfuRuntime runtime, UUID streamId) {
         }
     }
 
+    static AddressSpaceOptions addressSpaceOptions = new AddressSpaceOptions();
```
Comment (no2chem) [inline runtime/src/main/java/org/corfudb/recovery/RecoveryUtils.java]: Should be static final and be all caps and descriptive (something like DEFAULT_OPTIONS). Also It should be a member of AddressSpaceOptions

File: runtime/src/main/java/org/corfudb/runtime/CheckpointWriter.java [hunk 2]
```diff
@@ -127,7 +130,11 @@ public CheckpointWriter(CorfuRuntime rt, UUID streamId, String author, SMRMap ma
         this.map = map;
         checkpointId = UUID.randomUUID();
         checkpointStreamID = CorfuRuntime.getStreamID(streamId.toString() + "_cp");
-        sv = rt.getStreamsView();
+
+        StreamOptions options = StreamOptions.builder()
+                .doCache(false)
```
Comment (no2chem) [inline runtime/src/main/java/org/corfudb/runtime/CheckpointWriter.java]: Shouldn't this be setCached?

File: runtime/src/main/java/org/corfudb/runtime/view/AddressSpaceOptions.java [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+import lombok.Setter;
+
+/**
+ * Created by dmalkhi on 7/13/17.
+ */
+@AllArgsConstructor @NoArgsConstructor
+public class AddressSpaceOptions {
+    @Getter
+    @Setter
+    boolean doCache = true;
```
Comment (no2chem) [inline runtime/src/main/java/org/corfudb/runtime/view/AddressSpaceOptions.java]: should be called "cached"...

File: runtime/src/main/java/org/corfudb/runtime/view/AddressSpaceOptions.java [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+import lombok.AllArgsConstructor;
+import lombok.Getter;
+import lombok.NoArgsConstructor;
+import lombok.Setter;
+
+/**
+ * Created by dmalkhi on 7/13/17.
+ */
+@AllArgsConstructor @NoArgsConstructor
+public class AddressSpaceOptions {
```
Comment (no2chem) [inline runtime/src/main/java/org/corfudb/runtime/view/AddressSpaceOptions.java]: Needs comments

File: runtime/src/main/java/org/corfudb/runtime/view/AddressSpaceView.java [hunk 5]
```diff
@@ -120,7 +129,8 @@ public void write(IToken token, Object data) throws OverwriteException {
         });
 
         // Cache the successful write
-        if (!runtime.isCacheDisabled()) {
+        if ((opts == null || opts != null && opts.isDoCache() )
```
Comment (no2chem) [inline runtime/src/main/java/org/corfudb/runtime/view/AddressSpaceView.java]: simplify this conditional: ```!runtime.isCacheDisabled && (opts == null ||  opts.isCached())```
short-circuiting means second null check is not necessary

File: runtime/src/main/java/org/corfudb/runtime/view/AddressSpaceView.java [hunk 6]
```diff
@@ -147,7 +157,12 @@ public void write(IToken token, Object data) throws OverwriteException {
      * @return A result, which be cached.
      */
     public @Nonnull ILogData read(long address) {
-        if (!runtime.isCacheDisabled()) {
+        return read(address, null);
+    }
+
+    public @Nonnull ILogData read(long address, AddressSpaceOptions opts) {
+        if ((opts == null || opts != null && opts.isDoCache() )
```
Comment (no2chem) [inline runtime/src/main/java/org/corfudb/runtime/view/AddressSpaceView.java]: Same as above

File: runtime/src/main/java/org/corfudb/runtime/view/StreamOptions.java [hunk 7]
```diff
@@ -4,12 +4,14 @@
...
-    public static StreamOptions DEFAULT = new StreamOptions(false);
+    public static StreamOptions DEFAULT = new StreamOptions(false, true);
 
     public final boolean ignoreTrimmed;
+    public final boolean doCache;
 
-    public StreamOptions(boolean ignoreTrimmed) {
+    public StreamOptions(boolean ignoreTrimmed, boolean doCache) {
         this.ignoreTrimmed = ignoreTrimmed;
+        this.doCache = doCache;
```
Comment (no2chem) [inline runtime/src/main/java/org/corfudb/runtime/view/StreamOptions.java]: doCache = cached

File: runtime/src/main/java/org/corfudb/runtime/view/stream/BackpointerStreamView.java [hunk 8]
```diff
@@ -35,6 +39,7 @@
 public class BackpointerStreamView extends AbstractQueuedStreamView {
 
     final StreamOptions options;
+    final AddressSpaceOptions addressSpaceOptions = new AddressSpaceOptions();
```
Comment (no2chem) [inline runtime/src/main/java/org/corfudb/runtime/view/stream/BackpointerStreamView.java]: This really needs to be in AddressSpaceOptions and not all over each place where an addressspaceoption is needed

File: runtime/src/main/java/org/corfudb/runtime/view/AddressSpaceOptions.java (hunk 4 above)
Comment (medhavidhawan) [inline runtime/src/main/java/org/corfudb/runtime/view/AddressSpaceOptions.java]: I love this requirement from michael :)

**Comment by corfudb-performance** [comment]
Results automatically generated by _CorfuDB Benchmark Framework_ to assess the performance of this pull request for commit **a72e46e**.

*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 10 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Iterator workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Iterator workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Iterator workload, 10 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 10 threads, Disk mode
 
 An interactive dashboard with Pull Request Performance Metrics for ALL cluster types and numbers of threads in run, is available at:
 [Pull Request #869 Graphs ](https://snapshot.raintank.io/dashboard/snapshot/h18JfwkQtNtGsniHJhuRNoQRPsy4gZCD)

**Comment by corfudb-performance** [comment]
Results automatically generated by _CorfuDB Benchmark Framework_ to assess the performance of this pull request for commit **601595a**.

*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 10 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Iterator workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Iterator workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Iterator workload, 10 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 10 threads, Disk mode
 
 An interactive dashboard with Pull Request Performance Metrics for ALL cluster types and numbers of threads in run, is available at:
 [Pull Request #869 Graphs ](https://snapshot.raintank.io/dashboard/snapshot/wO4vAOXSineda7CjWUvFA7ekrjRW0FdD)

File: runtime/src/main/java/org/corfudb/runtime/view/AddressSpaceView.java [hunk 9]
```diff
@@ -185,6 +194,10 @@ public void write(IToken token, Object data) throws OverwriteException {
         return addressesMap;
     }
 
+    public Map<Long, ILogData> read(Iterable<Long> addresses) {
```
Comment (rogermichoud) [inline runtime/src/main/java/org/corfudb/runtime/view/AddressSpaceView.java]: This function does not guarantee ordering. The map returned by the cache is an ImmutableMap that references an HashMap. The HashMap is created in Caffeine code, so we have a bit our hands tight here. Maybe we can create a TreeMap out of the map returned by readCache.getAll ?

File: runtime/src/main/java/org/corfudb/runtime/view/StreamsView.java [hunk 10]
```diff
@@ -11,9 +11,7 @@
 import lombok.extern.slf4j.Slf4j;
 
 import org.corfudb.protocols.logprotocol.StreamCOWEntry;
-import org.corfudb.protocols.wireprotocol.TokenResponse;
-import org.corfudb.protocols.wireprotocol.TokenType;
-import org.corfudb.protocols.wireprotocol.TxResolutionInfo;
+import org.corfudb.protocols.wireprotocol.*;
```
Comment (rogermichoud) [inline runtime/src/main/java/org/corfudb/runtime/view/StreamsView.java]: Sneaky intellij "smart import" here. There is an option when you hover on the line to import all (and only) the used apis.

**Comment by dahliamalkhi** [comment]
don't merge yet.

**Comment by Maithem** [comment]
Addressed by #1330

**Pull request closed without merging**